internal/provider/resource_shuffle.go

```diff
@@ -95,71 +95,67 @@ func (r *shuffleResource) Schema(ctx context.Context, req resource.SchemaRequest
 }
 
 func (r *shuffleResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
-	var plan shuffleModelV0
-	diags := req.Plan.Get(ctx, &plan)
-	resp.Diagnostics.Append(diags...)
+	var data shuffleModelV0
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+
 	if resp.Diagnostics.HasError() {
 		return
 	}
 
-	input := plan.Input
-	seed := plan.Seed.ValueString()
-	resultCount := plan.ResultCount.ValueInt64()
+	// Legacy identifier attribute that is hardcoded. This is not necessary
+	// after Terraform 0.12, but left for compatibility reasons. The attribute
+	// could be removed in a future major version of the provider.
+	data.ID = types.StringValue("-")
 
-	if resultCount == 0 {
-		resultCount = int64(len(input.Elements()))
-	}
-
-	result := make([]attr.Value, 0, resultCount)
+	inputElements := data.Input.Elements()
 
-	if len(input.Elements()) > 0 {
-		rand := random.NewRand(seed)
+	var resultCount int64
 
-		// Keep producing permutations until we fill our result
-	Batches:
-		for {
-			perm := rand.Perm(len(input.Elements()))
+	if !data.ResultCount.IsNull() {
+		resultCount = data.ResultCount.ValueInt64()
+	} else {
+		resultCount = int64(len(inputElements))
+	}
 
-			for _, i := range perm {
-				result = append(result, input.Elements()[i])
+	// If the practitioner explicitly chose a result count of zero or the input
+	// had no elements, immediately return with an empty list for the result.
+	if resultCount == 0 || len(inputElements) == 0 {
+		data.Result = types.ListValueMust(types.StringType, []attr.Value{})
 
-				if int64(len(result)) >= resultCount {
-					break Batches
-				}
-			}
-		}
-	}
+		resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
 
-	res, diags := types.ListValue(types.StringType, result)
-	resp.Diagnostics.Append(diags...)
-	if resp.Diagnostics.HasError() {
 		return
 	}
 
-	s := shuffleModelV0{
-		ID:      types.StringValue("-"),
-		Keepers: plan.Keepers,
-		Input:   plan.Input,
-		Result:  res,
-	}
+	rand := random.NewRand(data.Seed.ValueString())
+	resultElements := make([]attr.Value, 0, resultCount)
 
-	if plan.Seed.IsNull() {
-		s.Seed = types.StringNull()
-	} else {
-		s.Seed = types.StringValue(seed)
-	}
+	// Keep producing permutations until we fill our result
+Batches:
```

Comment:
_A wild label break has appeared 🦀_

Reply:
😂 I _really wanted_ to refactor this out so bad, but tried to limit the changes to only those necessary.

Reply:
I like it. Adds character to the code 😆